c10/core/Stream.h

```diff
@@ -119,21 +119,21 @@ class Stream final {
     // that the bitmasking code below is updated accordingly!
     static_assert(sizeof(DeviceType) == 1, "DeviceType is not 8-bit");
     static_assert(sizeof(DeviceIndex) == 1, "DeviceIndex is not 8-bit");
-    static_assert(sizeof(StreamId) == 4, "DeviceIndex is not 32-bit");
+    static_assert(sizeof(StreamId) == 8, "DeviceIndex is not 64-bit");
     // Concat these together into a 64-bit integer
     // See Note [Hazard when concatenating signed integers]
     uint64_t bits = static_cast<uint64_t>(static_cast<uint8_t>(device_type()))
-            << 48 |
-        static_cast<uint64_t>(static_cast<uint8_t>(device_index())) << 32 |
+            << 56 |
+        static_cast<uint64_t>(static_cast<uint8_t>(device_index())) << 48 |
```

Comment:
It would be nice if there was an assert here that the bits we masked out were the signed extended highest order bit of the valid 48 bit segment, would make this code marginally more future proof.

Reply:
Actually, the code here is technically wrong. Because addresses are sign extended from the valid 48-bit range to 64-bit range, you are obligated to mask out the top 16 bits (and then when unpacking, sign extend the 48 bits). If you are given 0xFFFFFFFF FFFFFFFF (which is a valid address even under 48-bit assumption) you'd clobber the device type and device index here.

Reply:
lol, you are absolutely right and made me ashamed of myself 😂